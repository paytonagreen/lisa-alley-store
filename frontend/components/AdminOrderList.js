import React, { useState, useEffect } from "react";
import { useQuery, useMutation, gql } from "@apollo/client";
import OrderPagination from "./OrderPagination";
import Link from "next/link";
import Error from "./ErrorMessage";
import { formatDistance } from "date-fns";
import formatMoney from "../lib/formatMoney";
import AdminOrderStyles, { Title, ButtonDiv, FulfilledButton, OrderUL } from "./styles/AdminOrderStyles";
import { Center } from "./styles/ItemListStyles";
import { ordersPerPage } from "../config";
import FulfillmentToggler from './FulfillmentToggler';

const ALL_ORDERS_QUERY = gql`
  query ALL_ORDERS_QUERY($skip: Int = 0, $first: Int = ${ordersPerPage} ) {
    orders(skip: $skip, first: $first, orderBy: createdAt_DESC) {
      user {
        name
        email
        address1
        address2
        city
        state
        zip
      }
      id
      fulfilled
      total
      createdAt
      items {
        id
        title
        price
        description
        quantity
        image
      }
    }
  }
`;



const OrderList = ({ page }) => {
  const { data, error, loading } = useQuery(ALL_ORDERS_QUERY, {
    variables: {
      skip: ordersPerPage * page - ordersPerPage,
      first: ordersPerPage,
    }
  });
  
  if (loading) return <p>Loading...</p>;
  if (error) return <Error error={error} />;
  const { orders } = data;
  console.log(orders);
  if (data) {
    return (
      <div>
        <Center>
          <OrderPagination page={page}/>
        </Center>
          <OrderUL>
            {orders.map((order) => (
              <AdminOrderStyles key={order.id}>
                <ButtonDiv>
                  <Link
                    href={{
                      pathname: "/order",
                      query: { id: order.id },
                    }}
                  >
                    <a>
                      <div className="order-meta">
                        <p>
                          {formatDistance(
                            new Date(order.createdAt),
                            new Date()
                          )}{" "}
                          ago
                        </p>
                        <hr />
                        <h3>
                          <Title>
                            Order: (
                            {order.items.reduce((a, b) => a + b.quantity, 0)}{" "}
                            Items, {order.items.length} Products)
                          </Title>
                        </h3>
                        {order.items.map((item) => {
                          return (
                            <p key={item.title}>
                              {item.quantity} &times; {item.title}
                            </p>
                          );
                        })}
                        <hr />
                        <p>Total: {formatMoney(order.total)}</p>
                        <hr />
                        <p>
                          <Title>Name:</Title> {order.user.name}
                        </p>
                        <p>
                          <Title>Email:</Title> {order.user.email}
                        </p>
                        <p>
                          <Title>Address:</Title>
                        </p>
                        <p>{order.user.address1}</p>
                        {order.user.address2 && <p>{order.user.address2}</p>}
                        <p>
                          {order.user.city}, {order.user.state} {order.user.zip}
                        </p>
                      </div>
                    </a>
                  </Link>
                  <FulfillmentToggler order={order}/>
                </ButtonDiv>
              </AdminOrderStyles>
            ))}
          </OrderUL>
        <Center>
          <OrderPagination page={page} />
        </Center>
      </div>
    );
  }
};

export default OrderList;
