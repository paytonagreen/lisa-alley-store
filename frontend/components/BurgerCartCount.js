import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TransitionGroup, CSSTransition } from "react-transition-group";

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all .4s;
    backface-visibility: hidden;
  }
  /* Initial state of entered dot */
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

const Dot = styled.div`
  background: ${(props) => props.theme.yellow};
  color: ${props => props.theme.teal};
  border-radius: 50%;
  margin: auto;
  padding: 0.5rem;
  line-height: 2rem;
  max-width: 3rem;
  font-weight: 100;
  font-feature-settings: "tnum";
  font-variant-numeric: tabular-nums;
`;

const CartCount = ({ count }) => (
  <AnimationStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 400, exit: 400 }}
      >
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyles>
);

export default CartCount;