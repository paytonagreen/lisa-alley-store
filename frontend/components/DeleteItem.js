import React from "react";
import { useMutation, gql } from "@apollo/client";
import { ALL_ITEMS_QUERY } from "./Items";

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

const DeleteItem = ({ id, children }) => {
  function update(e, payload) {
    //manually update cache on client so it maches server
    // 1. Read cache for items that we want
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // 2. Filter deleted Item out
    data.items = data.items.filter(
      (item) => item.id !== payload.data.deleteItem.id
    );
    // 3. Put items back!
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  }
  const [deleteItem, { error }] = useMutation(DELETE_ITEM_MUTATION, {
    variables: { id },
    update: update,
  });
  return (
    <button
      onClick={() => {
        if (confirm("Are you sure you want to delete this item?")) {
          deleteItem().catch((err) => {
            alert(err.message);
          });
        }
      }}
    >
      {children}
    </button>
  );
};

export default DeleteItem;
