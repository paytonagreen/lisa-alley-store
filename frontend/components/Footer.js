import React from "react";
import FooterStyles from './styles/FooterStyles';
import Link from 'next/link';

const Footer = () => {
  return (
    <FooterStyles>
      <div className="social">
        <Link href="https://instagram.com/lalley">
          <a>
            <img src="../static/PNG/ig.png" alt="Instagram" />
          </a>
        </Link>
        <Link href="mailto: lisaalley@lisaalley.com">
          <a>
            <img src="../static/PNG/email.png" alt="Email" />
          </a>
        </Link>
        <Link href="https://facebook.com/lisa.alley">
          <a>
            <img src="../static/PNG/facebook3.png" alt="Email" />
          </a>
        </Link>
      </div>
      <div className="created">
        <a href="https://github.com/paytonagreen">
          Created By Payton Green &copy; 2020
        </a>
      </div>
    </FooterStyles>
  );
};

export default Footer;
