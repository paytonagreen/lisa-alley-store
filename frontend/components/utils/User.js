import { useQuery, gql } from '@apollo/client';
import propTypes from 'prop-types';

const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      email
      name
      permissions
      address1
      address2
      city
      state
      zip
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`;

const useUser = () => {
  const { data } = useQuery(CURRENT_USER_QUERY)
  if (data) {
    return data.me;
  }
}


export default useUser;
export {CURRENT_USER_QUERY, useUser};