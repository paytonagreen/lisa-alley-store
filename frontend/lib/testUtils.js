import casual from 'casual';
import React from 'react';
import { render } from '@testing-library/react';
import { ApolloProvider } from '@apollo/client';
import withData from '../lib/withData';

//setup wrapper for tests//
const Providers = ({ children, apollo }) => {
  return <ApolloProvider client={apollo} addTypename={false}>{children}</ApolloProvider>;
};

const customProviders = withData(Providers);

const customRender = (ui, options) =>
  render(ui, { wrapper: customProviders, ...options });

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
  createdAt: 1,
});

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  address1: casual.address1,
  address2: '',
  city: casual.city,
  state: casual.state,
  zip: 78909,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
});

const fakeRegularUser = () => ({
  __typename: 'User',
  id: '420',
  name: casual.name,
  email: casual.email,
  address1: casual.address,
  address2: '',

  city: casual.city,
  state: casual.state,
  zip: 11111,
  permissions: ['USER'],
  orders: [],
  cart: [],
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

const fakeCartUser = () => ({
  ...fakeUser(),
  cart: [fakeCartItem()],
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeRegularUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  fakeCartUser,
  customRender as render,
};
