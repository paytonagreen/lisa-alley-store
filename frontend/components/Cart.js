import React from "react";
import { useQuery, useMutation, gql } from "@apollo/client";
import { useUser } from "./User";
import CartStyles from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import SickButton from "./styles/SickButton";
import CartItem from "./CartItem";
import formatMoney from "../lib/formatMoney";
import calcTotalPrice from "../lib/calcTotalPrice";
import TakeMyMoney from "./TakeMyMoney";

const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Cart = () => {
  const me = useUser();
  const [toggleCart] = useMutation(TOGGLE_CART_MUTATION);
  const { data } = useQuery(LOCAL_STATE_QUERY);
  const cartOpen = data.cartOpen
  if (!me) return null;
  return (
    <CartStyles open={cartOpen}>
      <header>
        <CloseButton onClick={toggleCart} title="close">
          &times;
        </CloseButton>
        <Supreme>{me.name}'s Cart</Supreme>
        <p>
          You Have {me.cart.length} Item
          {me.cart.length === 1 ? "" : "s"} in your cart.
        </p>
      </header>
      <ul>
        {me.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalPrice(me.cart))}</p>
        {me.cart.length && (
          <TakeMyMoney>
            <SickButton onClick={() => toggleCart()}>Checkout</SickButton>
          </TakeMyMoney>
        )}
      </footer>
    </CartStyles>
  );
};

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };
