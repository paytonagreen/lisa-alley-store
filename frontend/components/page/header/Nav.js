import Link from "next/link";
import { useMutation } from "@apollo/client";
import NavStyles from "../../styles/NavStyles";
import { useUser } from "../../utils/User";
import Signout from "../../signup-signin/Signout";
import CartCount from "../../cart/CartCount";
import { TOGGLE_CART_MUTATION } from "../../cart/Cart";

function Nav() {
  const [toggleCart] = useMutation(TOGGLE_CART_MUTATION);
  const me = useUser();
  return (
    <NavStyles data-testid="nav">
      <Link href="/items">
        <a>Browse</a>
      </Link>
      {/* Signed In */}
      {me && (
        <>
          <Link href="/orders">
            <a>Orders</a>
          </Link>
          <Link href="/me">
            <a>Account</a>
          </Link>
          <Signout />
          <button onClick={toggleCart}>
            My Cart
            <CartCount
              count={me.cart.reduce(
                (tally, cartItem) => tally + cartItem.quantity,
                0
              )}
            />
          </button>
        </>
      )}
      {/* //Admin Only */}
      {me && me.permissions.includes("ADMIN") && (
        <Link href="/sell">
          <a>Sell</a>
        </Link>
      )}
      {/* //Not Signed In */}
      {!me && (
        <Link href="/signin">
          <a>Sign In</a>
        </Link>
      )}
    </NavStyles>
  );
}

export default Nav;
