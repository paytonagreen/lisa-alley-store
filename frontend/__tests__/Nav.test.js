import { mount } from "enzyme";
import wait from "waait";
import toJSON from "enzyme-to-json";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";
import { MockedProvider } from "@apollo/client/testing";
import { act } from "react-dom/test-utils";
import { fakeUser, fakeCartItem, fakeRegularUser } from "../lib/testUtils";

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeRegularUser() } },
  },
];

const signedInAdminMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } },
  },
];

const userWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
        },
      },
    },
  },
];

describe("<Nav/>", () => {
  let wrapper = ''
  it("renders a minimal nav when signed out", () => {
    act( () => {
      wrapper = mount(
        <MockedProvider mocks={notSignedInMocks}>
          <Nav />
        </MockedProvider>
      );
    });
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(toJSON(nav)).toMatchSnapshot();
  });

  it("renders full nav when signed in", async () => {
    await act(async () => {
      wrapper = mount(
        <MockedProvider mocks={signedInMocks}>
          <Nav />
        </MockedProvider>
      );
      await wait();
      wrapper.update();
    });
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(nav.children().length).toBe(5);
  });

  it("renders the sell button when ADMIN signed in", async () => {
    await act(async () => {
      wrapper = mount(
        <MockedProvider mocks={signedInAdminMocks}>
          <Nav />
        </MockedProvider>
      );
      await wait();
      wrapper.update();
    });
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(nav.children().length).toBe(6);
  });

  it("renders the amount of items in the cart", async () => {
    await act(async () => {
      wrapper = mount(
        <MockedProvider mocks={userWithCartItems}>
          <Nav />
        </MockedProvider>
      );
      await wait();
      wrapper.update();
    });
    const nav = wrapper.find('ul[data-test="nav"]');
    const count = nav.find("div.count");
    expect(toJSON(count)).toMatchSnapshot();
  });
});
