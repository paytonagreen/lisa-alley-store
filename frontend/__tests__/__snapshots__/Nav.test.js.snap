// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
<ul
  className="NavStyles-sc-11c0d2g-0 hfmIxm"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Browse
    </a>
  </Link>
  <Link
    href="/signin"
  >
    <a
      href="/signin"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Sign In
    </a>
  </Link>
</ul>
`;

exports[`<Nav/> renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 kDZyEW"
>
  9
</div>
`;
